pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -0,0 +1,366 @@
+	return vsphereServer.addresses
+}
+
+func (vsphereServer VSphereServer) Status() instance.Status {
+	return vsphereServer.status
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
+	return spec, false, nil
+}
+
+func getClient(username, password, address string, allowInsecure bool) (*govmomi.Client, error) {
+	clientUrl, err := url.Parse(fmt.Sprintf("%s/sdk", address))
+	if err != nil {
+		return nil, err
+	}
+	clientUrl.User = url.UserPassword(username, password)
+
+	return govmomi.NewClient(context.TODO(), clientUrl, allowInsecure)
+}
+
+func (p *provider) getConfig(s runtime.RawExtension) (*Config, *providerconfig.Config, error) {
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig)
+
+	c := Config{}
+	c.TemplateVMName, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TemplateVMName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Username, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Username)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Password, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Password)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.VSphereURL, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.VSphereURL)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Datacenter, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Datacenter)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Cluster, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Cluster)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.Datastore, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Datastore)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.AllowInsecure, err = p.configVarResolver.GetConfigVarBoolValue(rawConfig.AllowInsecure)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.CPUs = rawConfig.CPUs
+	c.MemoryMB = rawConfig.MemoryMB
+
+	return &c, &pconfig, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	config, _, err := p.getConfig(spec.ProviderConfig)
+	client, err := getClient(config.Username, config.Password, config.VSphereURL, config.AllowInsecure)
+	if err != nil {
+		return fmt.Errorf("failed to get vsphere client: '%v'", err)
+	}
+
+	finder, err := getDatacenterFinder(config.Datacenter, client)
+	if err != nil {
+		return err
+	}
+
+	_, err = finder.Datastore(context.TODO(), config.Datastore)
+	if err != nil {
+		return err
+	}
+
+	_, err = finder.ClusterComputeResource(context.TODO(), config.Cluster)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %v", err)
```

Comment:
some time ago we introduced the concept of `terminal` errors. Please change the following line to return a terminal error. This also applies to other places in the code.

examples:
https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L497
https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L608